onnx_chainer/replace_func.py

```diff
@@ -0,0 +1,129 @@
+import inspect
+from inspect import Parameter
+
+import chainer
+
+
+class WrappedFunctionNode(chainer.FunctionNode):
+
+    def __init__(self, name, func, args, kwargs, attributes=None):
+        self.user_name = name
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+
+        if attributes is not None:
+            self.set_attributes(attributes)
+
+    def set_attributes(self, attributes):
+        for p in attributes:
+            if isinstance(p, tuple):
+                assert isinstance(p[0], int)
+                setattr(self, p[1], self.args[p[0]])
+            elif isinstance(p, str):
+                setattr(self, p, self.kwargs.get(p, None))
+
+    def forward(self, xs):
+        self.xs = xs
+        results = self.func(*self.args, **self.kwargs)
+        if isinstance(results, (tuple, list)):
+            dummy_results = tuple(
+                ret.array if _isvar(ret) else ret for ret in results)
+        elif isinstance(results, dict):
+            dummy_results = tuple(
+                ret.array if _isvar(ret) else ret for ret in results.values())
+        else:
+            dummy_results = results.array if _isvar(results) else results
+            dummy_results = dummy_results,
+        return dummy_results
+
+    def backward(self, indexes, gys):
+        xp = chainer.backend.get_array_module(self.xs[0])
+        ret = tuple(chainer.Variable(xp.zeros_like(x)) for x in self.xs)
+        return ret
+
+
+def fake_as_funcnode(alt_func, name, attributes=None):
+    """The target function fakes FunctionNode
+
+    The target function is replaced to the alternative function to connect
+    variable node by acting function node. ``alt_func`` must satisfy the
+    following restrictions.
+
+    1. Inputs includes one or more ``chainer.Variable`` to trace variables.
```

Comment:
How about adding a bunch of error checks in `_wrapper` for these two restrictions?

Reply:
fixed, add error check on input and output